### ISSUE TITLE
[Question] entropy sign, once again

### ISSUE THREAD

**Issue by externalsupplierstaff** [DLR-RM/stable-baselines3#548]
### Question

I am confused with the entropy. The definition is $\text{entropy} =  - \sum_i p_i \log(p_i) $, but the code reads:

```python
                # Entropy loss favor exploration
                if entropy is None:
                    # Approximate entropy when no analytical form
                    entropy_loss = -th.mean(-log_prob)
                else:
                    entropy_loss = -th.mean(entropy)
```

and it gives the opposite result (in terms of concavity) than using entropy as in the definition (having the factor p_i before log(p_i)). I haven't read any papers, but I think the definition of entropy is quite well established. See different curve shapes

[wikipedia definition shape](https://www.wolframalpha.com/input/?i=-+x+log+x+-+%281-x%29+log+%281-x%29)
[ignoring p_i shape](https://www.wolframalpha.com/input/?i=-++log+x+-++log+%281-x%29)

what am I missing?

### Checklist

- [x ] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [ x] I have checked that there is ~no~ similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

**Comment by Miffyli** [comment]
If you are referring to the extra negation, this is because the variable `entropy_loss` is treated as a loss, i.e. something to be minimized (-> maximizing entropy). Note that `log_prob` is a vector of separate batch elements, so we estimate entropy for each individually (they are not meant to be summed over).

**Comment by externalsupplierstaff** [comment]
Hi @Miffyli , thanks for your kind answer. No, I'm not referring to the extra negation. I'll try to explain better. I think while answering I understood what the code is doing, but I'll like to write down the alternative.

I was referring to the second point that you make, about the vector character of `log_prob`. Is `log_prob` a matrix?
I thought that we needed, for each element in the batch, the full probability distribution over actions to compute its entropy. If `log_prob` is not a matrix (and I think it isn't, according to how is this variable used in the code), then what the code is doing is approximating the entropy by the logarithm of the probability of the selected (most likely) action. This makes sense, because roughly, a distribution with a higher probability in the top action has a higher entropy. Is this rationale right?

If it is, you can go ahead and close the topic. However, I think it would be nice to add to the comment "Approximate entropy when no analytical form" something like "as in _some reference". It took time to me to realize which was the approximation (and I don't even know if I got it right). Moreover, do you know the impact of using the full entropy definition vs this version?

Thanks

**Comment by Miffyli** [comment]
Ah, right! Thanks for the clarification!
Message got bit lengthy, but TL;DR: the estimate is definitely not the best one, but a) does not apply in many cases here and b) _might_ be good for RL (but definitely could be better!).

So for background, `actions` in [this loop](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/ppo/ppo.py#L189) (the one you pointed to) is a vector/matrix of actions of whatever type action_space defines. For discrete they would be
```python
# B x 1 matrix, where B is batch size
actions = [1, 2, 3, 0, ....]
```
The `log_probs` is a matrix (or rather, a vector), where it tells the probability of picking the corresponding action by the policy
```python
# B x 1 matrix (or a vector)
log_probs = [
    -2.41,
    -4.18,
    -4.61,
    -5.16,
    ...
]
```

Now, under the hood, [this function](https://github.com/DLR-RM/stable-baselines3/blob/3efab0d267e74cb03264411d4500ddde0c163404/stable_baselines3/common/policies.py#L630) is called to get the log_probs and entropies (if possible), which call the corresponding [distribution class](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/distributions.py). Most of the distributions have a proper entropy function for them which involves the whole distribution.

However distribution for [Squashed gaussian](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/distributions.py#L228) does not have a closed entropy, and since this was originally only proposed for SAC (which does not require entropy loss, as its whole point is embedding entropy in the objective function), there was no proper source on estimating it.

I guess the estimate `-log_prob` comes from taking the expectation-form of entropy `-E_x[log p(x)]` and substituting the expectation with a single-sample estimate, ending up with `-log p(x)`. This sounds odd tho because you are replacing one `p(x) = 1` (the one from expectation) but then leave `log p(x)` be the original value (??? :D). Seems like the "correct" estimation of entropy is a deep rabbit hole: https://math.stackexchange.com/questions/604654/estimating-the-entropy

For our RL purposes it _kind of_ is enough, because we simply want to discourage high-probability actions and encourage low-probability actions, which entropy _kind of_ does even in this sampling case (the gradient wants to take the probability where  [this](https://upload.wikimedia.org/wikipedia/commons/thumb/2/22/Binary_entropy_plot.svg/1200px-Binary_entropy_plot.svg.png) curve peaks). *However*, ideally entropy should encourage actions to be equally weighted, but if you have many options to pick from (e.g. think of big discrete distribution [for which we have true entropy tho]), then the "target probability" per action should be much lower than 0.5. I am not sure how this works with the continuous action spaces (e.g. the squashed gaussian), but I hazard to guess the same applies there.

**Comment by externalsupplierstaff** [comment]
Thanks very much for your detailed answers! Indeed, I missed that in the code we were using the entropy for each distribution, which makes much more sense, I thought we were using that estimate for my categorical distribution. That was my confusion. 

Regarding the current estimate... I don't know sac nor the area mentioned on the math exchange site. Definitely agree with your analysis of the current approach. I'm happy and will close this issue, but maybe a notimplementederror, a link to this issue or something else could be nice, as this is a bit obscure (I'm not certain yet that the estimator is working in the right direction, I think that the curve is inverted if we assume p=1) feel free to open again to discuss the point between parenthesis

**Comment by araffin** [comment]
As a side remark, we have a test that ensure this estimate is not too bad: https://github.com/DLR-RM/stable-baselines3/blob/master/tests/test_distributions.py#L79
(I think I took the idea from https://github.com/openai/baselines/blob/master/baselines/common/distributions.py#L323)
